src/fs.rs

```diff
@@ -0,0 +1,419 @@
+        // case right either).
+        if let Ok(md) = old_file.metadata() {
+            // TODO(MSRV): Consider replacing with std::os::unix::fs::fchown when MSRV >= 1.73
+            if unsafe { fchown(new_file.as_raw_fd(), md.uid(), md.gid()) } == -1 {
+                FLOG!(
+                    history_file,
+                    "Error when changing ownership of file:",
+                    errno::errno()
+                );
+            }
+            if let Err(e) = new_file.set_permissions(md.permissions()) {
+                FLOG!(history_file, "Error when changing mode of file:", e);
+            }
+        } else {
+            FLOG!(history_file, "Could not get metadata for file");
+        }
+        // Linux by default stores the mtime with low precision, low enough that updates that occur
+        // in quick succession may result in the same mtime (even the nanoseconds field). So
+        // manually set the mtime of the new file to a high-precision clock. Note that this is only
+        // necessary because Linux aggressively reuses inodes, causing the ABA problem; on other
+        // platforms we tend to notice the file has changed due to a different inode.
+        //
+        // The current time within the Linux kernel is cached, and generally only updated on a timer
+        // interrupt. So if the timer interrupt is running at 10 milliseconds, the cached time will
+        // only be updated once every 10 milliseconds.
+        #[cfg(any(target_os = "linux", target_os = "android"))]
+        {
+            let mut times: [libc::timespec; 2] = unsafe { std::mem::zeroed() };
+            times[0].tv_nsec = libc::UTIME_OMIT; // don't change atime
+            if unsafe { libc::clock_gettime(libc::CLOCK_REALTIME, &mut times[1]) } == 0 {
+                unsafe {
+                    // This accesses both times[0] and times[1]. Check `utimensat(2)` for details.
+                    libc::futimens(new_file.as_raw_fd(), &times[0]);
+                }
+            }
+        }
+    }
+
+    /// Renames a file from `old_name` to `new_name`.
+    fn rename(old_name: &wstr, new_name: &wstr) -> std::io::Result<()> {
+        if wrename(old_name, new_name) == -1 {
+            let error_number = errno::errno();
+            FLOG!(
+                error,
+                wgettext_fmt!("Error when renaming file: %s", error_number.to_string())
+            );
+            return Err(std::io::Error::from(error_number));
+        }
+        Ok(())
+    }
+
+    const TMP_FILE_SUFFIX: &wstr = L!(".XXXXXX");
+    let mut tmp_file_template = path.to_owned();
+    tmp_file_template.push_utfstr(TMP_FILE_SUFFIX);
+    let tmp_file_template = tmp_file_template;
+    let (tmp_file, tmp_name) = create_temporary_file(&tmp_file_template)?;
+    // Ensure that the temporary file is unlinked when this function returns.
+    let mut tmp_file = ScopeGuard::new(tmp_file, |_| {
+        wunlink(&tmp_name);
+    });
```

Comment:
great, I was gonna suggest that we extract all the `wunlink(&tmp_name);` added by the first few commits.